xds/internal/xdsclient/e2e_test/lds_watchers_test.go

```diff
@@ -0,0 +1,823 @@
+				ConfigType: &v3listenerpb.Filter_TypedConfig{TypedConfig: hcm},
+			}},
+		}},
+	}
+}
+
+// xdsClient is expected to produce an error containing this string when an
+// update is received containing a listener created using `badListenerResource`.
+const wantNACKErr = "no RouteSpecifier"
+
+// verifyNoListenerUpdate verifies that no listener update is received on the
+// provided update channel, and returns an error if an update is received.
+//
+// A very short deadline is used while waiting for the update, as this function
+// is intended to be used when an update is not expected.
+func verifyNoListenerUpdate(ctx context.Context, updateCh *testutils.Channel) error {
+	sCtx, sCancel := context.WithTimeout(ctx, defaultTestShortTimeout)
+	defer sCancel()
+	if u, err := updateCh.Receive(sCtx); err != context.DeadlineExceeded {
+		return fmt.Errorf("unexpected ListenerUpdate: %v", u)
+	}
+	return nil
+}
+
+// verifyListenerUpdate waits for an update to be received on the provided
+// update channel and verifies that it matches the expected update.
+//
+// Returns an error if no update is received before the context deadline expires
+// or the received update does not match the expected one.
+func verifyListenerUpdate(ctx context.Context, updateCh *testutils.Channel, wantUpdate xdsresource.ListenerUpdateErrTuple) error {
+	u, err := updateCh.Receive(ctx)
+	if err != nil {
+		return fmt.Errorf("timeout when waiting for a listener resource from the management server: %v", err)
+	}
+	got := u.(xdsresource.ListenerUpdateErrTuple)
+	if wantUpdate.Err != nil {
+		if gotType, wantType := xdsresource.ErrType(got.Err), xdsresource.ErrType(wantUpdate.Err); gotType != wantType {
+			return fmt.Errorf("received update with error type %v, want %v", gotType, wantType)
+		}
+	}
+	if diff := cmp.Diff(wantUpdate.Update, got.Update, cmpOpts...); diff != "" {
+		return fmt.Errorf("received unepected diff in the listener resource update: (-want, got):\n%s", diff)
+	}
+	return nil
+}
+
+// TestLDSWatch covers the case where a single watcher exists for a single
+// listener resource. The following sequence of events happen:
+// 1. A watch is registered for a listener resource
+// 2. Management server sends an update for the resource being watched
+// 3. Management server sends an update for a resource *not* being watched
+// 4. Watch is cancelled
+// 5. Management server sends an update for the original resource
+//
+// The test is run for old and new style names.
+func (s) TestLDSWatch(t *testing.T) {
+	overrideFedEnvVar(t)
+	mgmtServer, nodeID, bootstrapContents, _, cleanup := e2e.SetupManagementServer(t, nil)
+	defer cleanup()
+
+	// Create an xDS client with the above bootstrap contents.
+	client, err := xdsclient.NewWithBootstrapContentsForTesting(bootstrapContents)
+	if err != nil {
+		t.Fatalf("Failed to create xDS client: %v", err)
+	}
+	defer client.Close()
+
+	const (
+		ldsName = "xdsclient-test-lds-resource"
```

Comment:
Nit: These two consts are declared 7 times, taking up 14 lines. Make this a package scoped const?